### PR TITLE
fix: Ignore transient tap test directories

### LINKED ISSUE #348: [BUG] tap-testdir-* is not gitignored
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When running npm/cli tests on windows, I was seeing many files churn in and out of the source control files list.

These ephemeral test files slow down the IDE considerably and can make it hard to see what actually changed.

image
 

### Expected Behavior

Files created by tap tests should either be:
1. ignored by gitignore.
2. created outside the git working tree

### Steps To Reproduce

1. In VSCode on Windows, clone https://github.com/npm/cli
2. Open the Source Control pane
3. Run `npm test`

### Environment

- npm: 9.6.7
- Node: 18.17.1
- OS: Windows 11
- platform: AMD64 Desktop

### LINKED ISSUE #359: [BUG] lint script broken
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

`npm run lint`

```
> @npmcli/template-oss@4.18.0 lint
> eslint "**/*.js"


Oops! Something went wrong! :(

ESLint: 8.49.0

Error: ESLint configuration in test\apply\tap-testdir-files-snapshots-turn-off-specific-files\.eslintrc.yml is invalid:
        - Property "" is the wrong type (expected object but got `"exists"`).

    at ConfigValidator.validateConfigSchema (C:\Users\dan\Source\template-oss\node_modules\@eslint\eslintrc\dist\eslintrc.cjs:2168:19)
    at ConfigArrayFactory._normalizeConfigData (C:\Users\dan\Source\template-oss\node_modules\@eslint\eslintrc\dist\eslintrc.cjs:3010:19)
    at ConfigArrayFactory.loadInDirectory (C:\Users\dan\Source\template-oss\node_modules\@eslint\eslintrc\dist\eslintrc.cjs:2878:33)
    at CascadingConfigArrayFactory._loadConfigInAncestors (C:\Users\dan\Source\template-oss\node_modules\@eslint\eslintrc\dist\eslintrc.cjs:3860:46)
    at CascadingConfigArrayFactory.getConfigArrayForFile (C:\Users\dan\Source\template-oss\node_modules\@eslint\eslintrc\dist\eslintrc.cjs:3781:18)
    at FileEnumerator._iterateFilesRecursive (C:\Users\dan\Source\template-oss\node_modules\eslint\lib\cli-engine\file-enumerator.js:450:49)
    at _iterateFilesRecursive.next (<anonymous>)
    at FileEnumerator._iterateFilesRecursive (C:\Users\dan\Source\template-oss\node_modules\eslint\lib\cli-engine\file-enumerator.js:497:33)
    at _iterateFilesRecursive.next (<anonymous>)
    at FileEnumerator._iterateFilesRecursive (C:\Users\dan\Source\template-oss\node_modules\eslint\lib\cli-engine\file-enumerator.js:497:33)
```

### Expected Behavior

_No response_

### Steps To Reproduce

1. In this environment...
2. With this config...
3. Run '...'
4. See error...

### Environment

- npm:
- Node:
- OS: Windows 11
- platform:

### PR DESCRIPTION
**Pull request by rotu**
Running tests causes a lot of file churn. Although (usually) ephemeral, these files cause high CPU usage and can make developer tools difficult to use. These files also make linting fail.

Now the files will be ignored by git and by eslint.

## References
Fixes #348
Fixes #359

### CODE REVIEW

File: lib/content/index.js
Comment (wraithgar) [inline lib/content/index.js:158]: I'll let @lukekarrys weigh in here since it's surprising to me coming in fresh on this that we don't have anything in here already.

**Review by wraithgar — APPROVED** [review]

**Comment by wraithgar** [comment]
CI is goofy cause of a new `timers/promises` require, don't worry about it.

**Comment by rotu** [comment]
> CI is goofy cause of a new `timers/promises` require, don't worry about it.

That only explains the node@14 errors.

This is curious to me: https://github.com/npm/template-oss/actions/runs/6103503295/job/16623491287?pr=350

> Run npm i --prefer-online --no-fund --no-audit -g npm@latest
> npm ERR! code EBADENGINE
> npm ERR! engine Unsupported engine
> npm ERR! engine Not compatible with your version of node/npm: npm@10.0.0
> npm ERR! notsup Not compatible with your version of node/npm: npm@10.0.0
> npm ERR! notsup Required: {"node":"^18.17.0 || >=20.5.0"}
> npm ERR! notsup Actual:   {"npm":"8.6.0","node":"v18.0.0"}

I don't expect this since:

> Unless the user has set the `engine-strict` config flag, this field is advisory only and will only produce warnings when your package is installed as a dependency.

**Comment by wraithgar** [comment]
npm has a special check it does if it detects you are installing npm globally, strict engines checks are done then.  This started in npm 7

cf https://github.com/npm/cli/pull/3731

**Comment by wraithgar** [comment]
This is how we've been fixing the ci for packages that didn't get an engines bump to match npm 10 https://github.com/npm/read-package-json/pull/190

**Comment by rotu** [comment]
That's a major wtf! There should definitely more specific error, especially since the error is the exact same one packages can no longer enforce:
https://docs.npmjs.com/cli/v6/configuring-npm/package-json?v=true#enginestrict

**Comment by wraithgar** [comment]
Those are the v6 docs. npm 7 and above has this flag and obeys it if you specify it.

https://docs.npmjs.com/cli/v9/using-npm/config#engine-strict

**Comment by wraithgar** [comment]
Yeah after many rfc calls it was decided that package maintainers should not be able to specify things like that, only the consumers.

npm is a special snowflake because it IS what you use to install packages, so if it ends up installing a version that doesn't work in your current node, you can't run it to fix the problem.

File: lib/content/index.js
Comment (lukekarrys) [reply lib/content/index.js:158]: @rotu the current pattern is to put these paths in `lib/content/gitignore` and leave `ignorePaths` as a config to be set by consumers. can you make that change? apologies that this is not discoverable! 🙇🏼

**Comment by rotu** [comment]
Back to a draft to figure out interaction with https://github.com/tapjs/tapjs/commit/a5dc854b12ff9d8b3db7f623e041ac665260ca35

File: lib/content/index.js
Comment (rotu) [reply lib/content/index.js:158]: No worries. I can add the breadcrumbs that would have retroactively helped me. The `lib/content/gitignore` file is clearly a template of some sort. What instantiates this template?

Comment (rotu) [reply lib/content/index.js:158]: Also, any reason why `allowPaths` and `distPaths` are inlined in this file but `ignorePaths` is not?

File: lib/content/index.js
Comment (wraithgar) [inline lib/content/index.js:159]: Very helpful thank you.

File: package.json
Comment (wraithgar) [inline package.json:26]: This will need to be reverted

File: workspace/test-workspace/package.json
Comment (wraithgar) [inline workspace/test-workspace/package.json:23]: This will need to be reverted.

**Review by wraithgar — CHANGES_REQUESTED** [review]

File: package.json
Comment (rotu) [reply package.json:26]: Thanks! root cause: #363

**Review by lukekarrys — APPROVED** [review]

**Review by lukekarrys — APPROVED** [review]

**Review by wraithgar — APPROVED** [review]
*sigh* I'm gonna miss using `git status` to find where `TAP_SAVE_FIXTURE` left its stuff for a given test but this is the right choice.

**Comment by wraithgar** [comment]
@rotu I can't land this cause of conflicts, and they're the kind I can't even resolve in the GitHub ui :/

**Comment by lukekarrys** [comment]
@rotu can you try fetching the latest `origin/main` and then running `git rebase origin/main` on your feature branch? That's what I normally do when conflicts get in a weird state.

**Review by wraithgar — APPROVED** [review]

**Comment by wraithgar** [comment]
Rebased in https://github.com/npm/template-oss/pull/364 once @lukekarrys approves it'll auto merge.

**Comment by wraithgar** [comment]
Sorry about the merge conflicts @rotu, we got there in the end.  #364 landed.

Thanks again for all the help.

**Comment by rotu** [comment]
Oops! I was going to do that, but thank you for the favor, @wraithgar!

**Pull request closed without merging**
